hal/armv7a/exceptions.c

```diff
@@ -0,0 +1,138 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system loader
+ *
+ * Exception handling
+ *
+ * Copyright 2017, 2018, 2021 Phoenix Systems
+ * Author: Pawel Pisarczyk, Jakub Sejdak, Aleksander Kaminski, Hubert Buczynski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <hal/hal.h>
+
+
+typedef struct _exc_context_t {
+	u32 savesp;
+
+	u32 dfsr;
+	u32 dfar;
+	u32 ifsr;
+	u32 ifar;
+
+	u32 psr;
+
+	u32 r0;
+	u32 r1;
+	u32 r2;
+	u32 r3;
+	u32 r4;
+	u32 r5;
+	u32 r6;
+	u32 r7;
+	u32 r8;
+	u32 r9;
+	u32 r10;
+
+	u32 fp;
+	u32 ip;
+	u32 sp;
+	u32 lr;
+
+	u32 pc;
+} exc_context_t;
+
+
+static const char digits[] = "0123456789abcdef";
+
+
+static int exceptions_i2s(char *prefix, char *s, unsigned int i, unsigned char b, char zero)
+{
+	char c;
+	unsigned int l, k, m;
+
+	m = hal_strlen(prefix);
+	hal_memcpy(s, prefix, m);
+
+	for (k = m, l = (unsigned int)-1; l; i /= b, l /= b) {
+		if (!zero && !i)
+			break;
+		s[k++] = digits[i % b];
+	}
+
+	l = k--;
+
+	while (k > m) {
+		c = s[m];
+		s[m++] = s[k];
+		s[k--] = c;
+	}
+
+	return l;
+}
+
+
+void hal_exceptionsDumpContext(char *buff, exc_context_t *ctx, int n)
+{
+	static const char *const mnemonics[] = {
+		"0 #Reset", "1 #Undef", "2 #Syscall", "3 #Prefetch",
+		"4 #Abort", "5 #Reserved", "6 #FIRQ", "7 #IRQ"
+	};
+	size_t i = 0;
+
+	n &= 0x7;
+
+	hal_strcpy(buff, "\nException: ");
+	hal_strcpy(buff += hal_strlen(buff), mnemonics[n]);
+	hal_strcpy(buff += hal_strlen(buff), "\n");
+	buff += hal_strlen(buff);
+
+	i += exceptions_i2s(" r0=", &buff[i], ctx->r0, 16, 1);
```

Comment:
```suggestion
	hal_strcpy(buff, "\nException: ");
	hal_strcpy(buff += hal_strlen(buff), mnemonics[n]);
	buff += hal_strlen(buff);

	i += exceptions_i2s("\n r0=", &buff[i], ctx->r0, 16, 1);
```
will take less bytes

Because hal_strlen() and hal_strcpy() are defined as inline, you may define e.g. funcyion like this:
```
static char *ex_copy(char *s1, const char *s2)
{
	while ((*s1++ = *s2++))
		;
	s1--;

	return s1;
}```

and use it instead of those functions:
```
	buff = ex_copy(buff, "\nException: ");
	buff = ex_copy(buff, mnemonics[n]);
```

Reply:
Functions in `string.h` are no longer defined as inline so I think this code can be left as it is.